File: core/src/main/java/org/apache/iceberg/DataFiles.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;
import org.apache.hadoop.fs.FileStatus;
import org.apache.iceberg.encryption.EncryptedOutputFile;
import org.apache.iceberg.encryption.EncryptionKeyMetadata;
import org.apache.iceberg.hadoop.HadoopInputFile;
import org.apache.iceberg.io.InputFile;
import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
import org.apache.iceberg.types.Conversions;
import org.apache.iceberg.util.ByteBuffers;

public class DataFiles {

  private DataFiles() {}

  static PartitionData newPartitionData(PartitionSpec spec) {
    return new PartitionData(spec.partitionType());
  }

  static PartitionData copyPartitionData(
      PartitionSpec spec, StructLike partitionData, PartitionData reuse) {
    Preconditions.checkArgument(
        spec.isPartitioned(), "Can't copy partition data to a unpartitioned table");
    PartitionData data = reuse;
    if (data == null) {
      data = newPartitionData(spec);
    }

    Class<?>[] javaClasses = spec.javaClasses();
    List<PartitionField> fields = spec.fields();
    for (int i = 0; i < fields.size(); i += 1) {
      data.set(i, partitionData.get(i, javaClasses[i]));
    }

    return data;
  }

  static PartitionData fillFromPath(PartitionSpec spec, String partitionPath, PartitionData reuse) {
    PartitionData data = reuse;
    if (data == null) {
      data = newPartitionData(spec);
    }

    String[] partitions = partitionPath.split("/", -1);
    Preconditions.checkArgument(
        partitions.length <= spec.fields().size(),
        "Invalid partition data, too many fields (expecting %s): %s",
        spec.fields().size(),
        partitionPath);
    Preconditions.checkArgument(
        partitions.length >= spec.fields().size(),
        "Invalid partition data, not enough fields (expecting %s): %s",
        spec.fields().size(),
        partitionPath);

    for (int i = 0; i < partitions.length; i += 1) {
      PartitionField field = spec.fields().get(i);
      String[] parts = partitions[i].split("=", 2);
      Preconditions.checkArgument(
          parts.length == 2 && parts[0] != null,
          "Invalid partition: %s",
          partitions[i]);
      Preconditions.checkArgument(
          field.name().equals(parts[0]),
          "Invalid partition key (expecting %s): %s",
          field.name(),
          parts[0]);

      data.set(i, Conversions.fromPartitionString(data.getType(i), parts[1]));
    }

    return data;
  }

  static PartitionData fillFromValues(
      PartitionSpec spec, List<String> partitionValues, PartitionData reuse) {
    PartitionData data = reuse;
    if (data == null) {
      data = newPartitionData(spec);
    }

    Preconditions.checkArgument(
        partitionValues.size() == spec.fields().size(),
        "Invalid partition data, expecting %s fields, found %s",
        spec.fields().size(),
        partitionValues.size());

    for (int i = 0; i < partitionValues.size(); i += 1) {
      data.set(i, Conversions.fromPartitionString(data.getType(i), partitionValues.get(i)));
    }

    return data;
  }

  public static PartitionData data(PartitionSpec spec, String partitionPath) {
    return fillFromPath(spec, partitionPath, null);
  }

  public static PartitionData copy(PartitionSpec spec, StructLike partition) {
    return copyPartitionData(spec, partition, null);
  }

  public static DataFile fromManifest(ManifestFile manifest) {
    Preconditions.checkArgument(
        manifest.addedFilesCount() != null && manifest.existingFilesCount() != null,
        "Cannot create data file from manifest: data file counts are missing.");

    return DataFiles.builder(PartitionSpec.unpartitioned())
        .withPath(manifest.path())
        .withFormat(FileFormat.AVRO)
        .withRecordCount(manifest.addedFilesCount() + manifest.existingFilesCount())
        .withFileSizeInBytes(manifest.length())
        .build();
  }

  public static Builder builder(PartitionSpec spec) {
    return new Builder(spec);
  }

  public static class Builder {
    private final PartitionSpec spec;
    private final boolean isPartitioned;
    private final int specId;
    private PartitionData partitionData;
    private String filePath = null;
    private FileFormat format = null;
    private long recordCount = -1L;
    private long fileSizeInBytes = -1L;

    // optional fields
    private Map<Integer, Long> columnSizes = null;
    private Map<Integer, Long> valueCounts = null;
    private Map<Integer, Long> nullValueCounts = null;
    private Map<Integer, Long> nanValueCounts = null;
    private Map<Integer, ByteBuffer> lowerBounds = null;
    private Map<Integer, ByteBuffer> upperBounds = null;
    private ByteBuffer keyMetadata = null;
    private List<Long> splitOffsets = null;
    private Integer sortOrderId = SortOrder.unsorted().orderId();

    public Builder(PartitionSpec spec) {
      this.spec = spec;
      this.specId = spec.specId();
      this.isPartitioned = spec.isPartitioned();
      this.partitionData = isPartitioned ? newPartitionData(spec) : null;
    }

    public void clear() {
      if (isPartitioned) {
        partitionData.clear();
      }
      this.filePath = null;
      this.format = null;
      this.recordCount = -1L;
      this.fileSizeInBytes = -1L;
      this.columnSizes = null;
      this.valueCounts = null;
      this.nullValueCounts = null;
      this.nanValueCounts = null;
      this.lowerBounds = null;
      this.upperBounds = null;
      this.splitOffsets = null;
      this.sortOrderId = SortOrder.unsorted().orderId();
    }

    public Builder copy(DataFile toCopy) {
      if (isPartitioned) {
        Preconditions.checkState(
            specId == toCopy.specId(), "Cannot copy a DataFile with a different spec");
        this.partitionData = copyPartitionData(spec, toCopy.partition(), partitionData);
      }
      this.filePath = toCopy.path().toString();
      this.format = toCopy.format();
      this.recordCount = toCopy.recordCount();
      this.fileSizeInBytes = toCopy.fileSizeInBytes();
      this.columnSizes = toCopy.columnSizes();
      this.valueCounts = toCopy.valueCounts();
      this.nullValueCounts = toCopy.nullValueCounts();
      this.nanValueCounts = toCopy.nanValueCounts();
      this.lowerBounds = toCopy.lowerBounds();
      this.upperBounds = toCopy.upperBounds();
      this.keyMetadata =
          toCopy.keyMetadata() == null ? null : ByteBuffers.copy(toCopy.keyMetadata());
      this.splitOffsets =
          toCopy.splitOffsets() == null ? null : ImmutableList.copyOf(toCopy.splitOffsets());
      this.sortOrderId = toCopy.sortOrderId();
      return this;
    }

    public Builder withStatus(FileStatus stat) {
      this.filePath = stat.getPath().toString();
      this.fileSizeInBytes = stat.getLen();
      return this;
    }

    public Builder withInputFile(InputFile file) {
      if (file instanceof HadoopInputFile) {
        return withStatus(((HadoopInputFile) file).getStat());
      }

      this.filePath = file.location();
      this.fileSizeInBytes = file.getLength();
      return this;
    }

    public Builder withEncryptedOutputFile(EncryptedOutputFile newEncryptedFile) {
      withInputFile(newEncryptedFile.encryptingOutputFile().toInputFile());
      withEncryptionKeyMetadata(newEncryptedFile.keyMetadata());
      return this;
    }

    public Builder withPath(String newFilePath) {
      this.filePath = newFilePath;
      return this;
    }

    public Builder withFormat(String newFormat) {
      this.format = FileFormat.fromString(newFormat);
      return this;
    }

    public Builder withFormat(FileFormat newFormat) {
      this.format = newFormat;
      return this;
    }

    public Builder withPartition(StructLike newPartition) {
      if (isPartitioned) {
        this.partitionData = copyPartitionData(spec, newPartition, partitionData);
      }
      return this;
    }

    public Builder withRecordCount(long newRecordCount) {
      this.recordCount = newRecordCount;
      return this;
    }

    public Builder withFileSizeInBytes(long newFileSizeInBytes) {
      this.fileSizeInBytes = newFileSizeInBytes;
      return this;
    }

    public Builder withPartitionPath(String newPartitionPath) {
      Preconditions.checkArgument(
          isPartitioned || newPartitionPath.isEmpty(),
          "Cannot add partition data for an unpartitioned table");
      if (!newPartitionPath.isEmpty()) {
        this.partitionData = fillFromPath(spec, newPartitionPath, partitionData);
      }
      return this;
    }

    public Builder withPartitionValues(List<String> partitionValues) {
      Preconditions.checkArgument(
          isPartitioned ^ partitionValues.isEmpty(),
          "Table must be partitioned or partition values must be empty");
      if (!partitionValues.isEmpty()) {
        this.partitionData = fillFromValues(spec, partitionValues, partitionData);
      }
      return this;
    }

    public Builder withMetrics(Metrics metrics) {
      // check for null to avoid NPE when unboxing
      this.recordCount = metrics.recordCount() == null ? -1 : metrics.recordCount();
      this.columnSizes = metrics.columnSizes();
      this.valueCounts = metrics.valueCounts();
      this.nullValueCounts = metrics.nullValueCounts();
      this.nanValueCounts = metrics.nanValueCounts();
      this.lowerBounds = metrics.lowerBounds();
      this.upperBounds = metrics.upperBounds();
      return this;
    }

    public Builder withSplitOffsets(List<Long> offsets) {
      if (offsets != null) {
        this.splitOffsets = ImmutableList.copyOf(offsets);
      } else {
        this.splitOffsets = null;
      }
      return this;
    }

    /** @deprecated since 1.5.0, will be removed in 1.6.0; must not be set for data files. */
    @Deprecated
    public Builder withEqualityFieldIds(List<Integer> equalityIds) {
      throw new UnsupportedOperationException("Equality field IDs must not be set for data files");
    }

    public Builder withEncryptionKeyMetadata(ByteBuffer newKeyMetadata) {
      this.keyMetadata = newKeyMetadata;
      return this;
    }

    public Builder withEncryptionKeyMetadata(EncryptionKeyMetadata newKeyMetadata) {
      return withEncryptionKeyMetadata(newKeyMetadata.buffer());
    }

    public Builder withSortOrder(SortOrder newSortOrder) {
      if (newSortOrder != null) {
        this.sortOrderId = newSortOrder.orderId();
      }
      return this;
    }

    public DataFile build() {
      Preconditions.checkArgument(filePath != null, "File path is required");
      if (format == null) {
        this.format = FileFormat.fromFileName(filePath);
      }
      Preconditions.checkArgument(format != null, "File format is required");
      Preconditions.checkArgument(fileSizeInBytes >= 0, "File size is required");
      Preconditions.checkArgument(recordCount >= 0, "Record count is required");

      return new GenericDataFile(
          specId,
          filePath,
          format,
          isPartitioned ? partitionData.copy() : null,
          fileSizeInBytes,
          new Metrics(
              recordCount,
              columnSizes,
              valueCounts,
              nullValueCounts,
              nanValueCounts,
              lowerBounds,
              upperBounds),
          keyMetadata,
          splitOffsets,
          sortOrderId);
    }
  }
}

File: python/iceberg/core/data_files.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


from iceberg.api import (FileFormat,
                         Metrics)
from iceberg.api.types import Conversions

from .generic_data_file import GenericDataFile
from .partition_data import PartitionData


class DataFiles(object):
    DEFAULT_BLOCK_SIZE = 64 * 1024 * 1024

    @staticmethod
    def new_partition_data(spec):
        return PartitionData(spec.partition_type())

    @staticmethod
    def copy_partition_data(spec, partition_data, reuse):
        data = reuse
        if data is None:
            data = DataFiles.new_partition_data(spec)

        for i, _ in enumerate(spec.fields):
            data.set(i, partition_data.get(i))

        return data

    @staticmethod
    def fill_from_path(spec, partition_path, reuse):
        data = reuse
        if data is None:
            data = DataFiles.new_partition_data(spec)

        partitions = partition_path.split("/")
        if len(partitions) > len(spec.fields):
            raise RuntimeError("Invalid partition data, too many fields (expecting %s): %s" % (len(spec.fields),
                                                                                               partition_path))

        if len(partitions) < len(spec.fields):
            raise RuntimeError("Invalid partition data, not enough fields(expecting %s): %s" % (len(spec.fields),
                                                                                                partition_path))

        for i, part in enumerate(partitions):
            field = spec.fields[i]
            parts = part.split("=")
            if len(parts) != 2 or parts[0] is None or parts[0] != field.name:
                raise RuntimeError("Invalid partition: %s" % part)

            data.set(i, Conversions.from_partition_string(data.get_type(i), parts[1]))

        return data

    @staticmethod
    def data(spec, partition_path):
        return DataFiles.fill_from_path(spec, partition_path, None)

    @staticmethod
    def copy(spec, partition):
        return DataFiles.copy_partition_data(spec, partition, None)

    @staticmethod
    def from_input_file(input_file, row_count, partition_data=None, metrics=None):
        from .filesystem import FileSystemInputFile
        if isinstance(input_file, FileSystemInputFile):
            return DataFiles.from_stat(input_file.get_stat(), row_count,
                                       partition_data=partition_data, metrics=metrics)

    @staticmethod
    def from_stat(stat, row_count, partition_data=None, metrics=None):
        location = stat.path
        format = FileFormat.from_file_name(location)
        return GenericDataFile(location, format, stat.length, stat.block_size,
                               row_count=row_count, partition=partition_data, metrics=metrics)

    @staticmethod
    def builder(spec=None):
        return DataFileBuilder(spec)


class DataFileBuilder(object):

    def __init__(self, spec=None):
        self.spec = spec
        self.partition_data = DataFiles.new_partition_data(spec) if spec is not None else None
        self.is_partitioned = spec is not None
        self.file_path = None
        self.format = None
        self.record_count = -1
        self.file_size_in_bytes = -1
        self.block_size_in_bytes = -1
        self.column_sizes = None
        self.value_counts = None
        self.null_value_counts = None
        self.lower_bounds = None
        self.upper_bounds = None

    def clear(self):
        if self.is_partitioned:
            self.partition_data.clear()
        self.file_path = None
        self.format = None
        self.record_count = -1
        self.file_size_in_bytes = -1
        self.block_size_in_bytes = -1
        self.column_sizes = None
        self.value_counts = None
        self.null_value_counts = None
        self.lower_bounds = None
        self.upper_bounds = None
        return self

    def copy(self, to_copy):
        if self.is_partitioned:
            self.partition_data = DataFiles.copy_partition_data(self.spec, to_copy.partition, self.partition_data)

        self.file_path = str(to_copy.path())
        self.format = to_copy.format
        self.record_count = to_copy.record_count
        self.file_size_in_bytes = to_copy.file_size_in_bytes
        self.block_size_in_bytes = to_copy.block_size_in_bytes
        self.column_sizes = to_copy.column_sizes
        self.value_counts = to_copy.value_counts
        self.null_value_counts = to_copy.null_value_counts
        self.lower_bounds = to_copy.lower_bounds
        self.upper_bounds = to_copy.upper_bounds
        return self

    def with_status(self, stat):
        self.file_path = stat.path
        self.file_size_in_bytes = stat.length
        self.block_size_in_bytes = stat.blocksize
        return self

    def with_input_file(self, input_file):
        from .filesystem import FileSystemInputFile
        if isinstance(input_file, FileSystemInputFile):
            self.with_status(input_file.get_stat())

        self.file_path = self.location()
        self.file_size_in_bytes = self.get_length()

        return self

    def with_path(self, path):
        self.file_path = path
        return self

    def with_format(self, fmt):
        if isinstance(fmt, FileFormat):
            self.format = fmt
        else:
            self.format = FileFormat[str(fmt).upper()]

        return self

    def with_partition(self, partition):
        self.partition_data = DataFiles.copy_partition_data(self.spec, partition, self.partition_data)
        return self

    def with_record_count(self, record_count):
        self.record_count = record_count
        return self

    def with_file_size_in_bytes(self, file_size_in_bytes):
        self.file_size_in_bytes = file_size_in_bytes
        return self

    def with_block_size_in_bytes(self, block_size_in_bytes):
        self.block_size_in_bytes = block_size_in_bytes
        return self

    def with_partition_path(self, partition_path):
        if not self.is_partitioned:
            raise RuntimeError("Cannot add partition data for an unpartitioned table")

        self.partition_data = DataFiles.fill_from_path(self.spec, partition_path, self.partition_data)
        return self

    def with_metrics(self, metrics):
        self.record_count = metrics.row_count if metrics.row_count is not None else -1
        self.column_sizes = metrics.column_sizes
        self.value_counts = metrics.value_counts
        self.null_value_counts = metrics.null_value_counts
        self.lower_bounds = metrics.lower_bounds
        self.upper_bounds = metrics.upper_bounds
        return self

    def build(self):
        if self.file_path is None:
            raise RuntimeError("File path is required")
        if self.format is None:
            self.format = FileFormat.from_file_name(self.file_path)

        if self.format is None:
            raise RuntimeError("File format is required")

        if self.file_size_in_bytes < 0:
            raise RuntimeError("File size is required")

        if self.record_count < 0:
            raise RuntimeError("Record count is required")

        if self.block_size_in_bytes is None:
            self.block_size_in_bytes = DataFiles.DEFAULT_BLOCK_SIZE

        return GenericDataFile(self.file_path, self.format, self.file_size_in_bytes, self.block_size_in_bytes,
                               partition=self.partition_data.copy() if self.is_partitioned else None,
                               metrics=Metrics(row_count=self.record_count, column_sizes=self.column_sizes,
                                               value_counts=self.value_counts,
                                               null_value_counts=self.null_value_counts,
                                               lower_bounds=self.lower_bounds, upper_bounds=self.upper_bounds))
